### PR TITLE
Add connection warmup handler

### PR DESCRIPTION
**Pull request by shangchengying**
Add a handler for handling warm up connections to peer clusters after watchers are established.

Since establishing connections can be expensive and all the clients need to connect to hosts in the downstream's peer clusters, by pre-establishing the connections, we can avoid the thundering herd effect of all clients trying to connect at the same time.

### CODE REVIEW

File: d2/src/main/java/com/linkedin/d2/balancer/clusterfailout/FailedoutClusterConnectionWarmUpHandler.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package com.linkedin.d2.balancer.clusterfailout;
```
Comment (jpstewart) [inline d2/src/main/java/com/linkedin/d2/balancer/clusterfailout/FailedoutClusterConnectionWarmUpHandler.java:1]: License header

**Review by jpstewart — APPROVED** [review]

File: d2/src/main/java/com/linkedin/d2/balancer/clusterfailout/ZKFailoutConfigProvider.java [hunk 2]
```diff
@@ -100,4 +101,10 @@ public void onClusterRemoved(String clusterName)
       manager.updateFailoutConfig(null);
     }
   }
+
+  @Nullable
+  public FailedoutClusterConnectionWarmUpHandler createConnectionWarmUpHandler()
+  {
+    return null;
```
Comment (bbarkley) [inline d2/src/main/java/com/linkedin/d2/balancer/clusterfailout/ZKFailoutConfigProvider.java:108]: Might want to indicate this is meant to be overridden by subclasses

**Review by bbarkley — APPROVED** [review]
Suggestion for clearing up some code that isn't obvious, but otherwise looks good.

File: d2/src/main/java/com/linkedin/d2/balancer/clusterfailout/ZKFailoutConfigProvider.java (hunk 2 above)
Comment (shangchengying) [reply d2/src/main/java/com/linkedin/d2/balancer/clusterfailout/ZKFailoutConfigProvider.java:118]: Forgot to add docstring. Fixed

**Pull request merged**